stout/flags.cc

```diff
@@ -18,65 +18,204 @@ namespace stout::flags {
+  // names(stout.v1.flag->names: ["..."]).
+  std::set<std::string> flag_names;
+
   for (int i = 0; i < descriptor->field_count(); i++) {
     const auto* field = descriptor->field(i);
 
-    const auto& flag = field->options().GetExtension(stout::v1::flag);
+    // We need this descriptor for the subcommand's logic.
+    const google::protobuf::OneofDescriptor* oneof =
+        field->real_containing_oneof();
+
+    // Check if the current field of the message is in 'oneof'.
+    if (oneof != nullptr) {
+      // Check if the name of 'oneof' is 'subcommand'. If no -> exit with
+      // an error.
+      if (oneof->name() != "subcommand") {
+        std::cerr << "'oneof' field must have 'subcommand' name. "
+                  << "Other names are illegal"
+                  << std::endl;
+        exit(1);
+      } else {
+        // Subcommands must have stout.v1.subcommand option.
+        if (!field->options().HasExtension(stout::v1::subcommand)) {
+          std::cerr << "Every field of the 'oneof subcommand' must "
+                       "be annotated with a stout.v1.subcommand option"
+                    << std::endl;
+          exit(1);
+        } else {
+          // Check for missing 'names' and 'help'.
+          const auto& subcommand =
+              field->options().GetExtension(stout::v1::subcommand);
+
+          if (subcommand.names().empty()) {
+            std::cerr
+                << "Missing at least one subcommand name in 'names' "
+                << "for field '" << field->full_name() << "'"
+                << std::endl;
+            std::exit(1);
+          }
 
-    if (flag.names().empty()) {
-      std::cerr
-          << "Missing at least one flag name in 'names' for field '"
-          << field->full_name() << "'"
-          << std::endl;
-      std::exit(1);
-    }
+          if (subcommand.help().empty()) {
+            std::cerr
+                << "Missing subcommand 'help' for field '"
+                << field->full_name() << "'"
+                << std::endl;
+            std::exit(1);
+          }
+
+          // Helper for collecting all possible names for a subcommand,
+          // including deprecated names.
+          std::set<std::string> subcommand_names;
+
+          for (const auto& name : subcommand.names()) {
+            if (auto iterator =
+                    std::find_if(
+                        nested_parsers_.begin(),
+                        nested_parsers_.end(),
+                        [&name](const auto& parser) {
+                          return parser.second.count(name) > 0;
+                        });
+                iterator != nested_parsers_.end()) {
+              std::cerr << "Encountered duplicate subcommand name '"
+                        << name << "' for message '"
+                        << message->GetTypeName() << "'"
+                        << std::endl;
+              std::exit(1);
+            }
+            subcommand_fields_.emplace(name, field);
```

Comment:
seems that you store the same value `field` with different keys `name`, it is redundant there, we can avoid it using `std::map<field, std::vector<std::string>>` at least or you can create your own container that stores `field` ptr and `std::set` for names and override `operator[]` for your needs. Let's chat about it.